File: Python solutions/160. Intersection of Two Linked Lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None
class Solution:
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> Optional[ListNode]:
        def getLen(node):
            length = 0
            while node:
                length += 1
                node = node.next
            return length

        lenA = getLen(headA)
        lenB = getLen(headB)
        
        while lenA > lenB:
            lenA -= 1
            headA = headA.next
        
        while lenA < lenB:
            lenB -= 1
            headB = headB.next

        while headA != headB:
            headA = headA.next
            headB = headB.next
        
        return headA

File: C++ solutions/160. Intersection of Two Linked Lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(nullptr) {}
 * };
 */
class Solution {
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {

        auto getLen = [](ListNode *node) {
            int length = 0;
            while (node) {
                length++;
                node = node->next;
            }
            return length;
        };

        int lenA = getLen(headA);
        int lenB = getLen(headB);
        
        while (lenA > lenB) {
            lenA--;
            headA = headA->next;
        }
        while (lenB > lenA) {
            lenB--;
            headB = headB->next;
        }

        while (headA != headB) {
            headA = headA->next;
            headB = headB->next;
        }
        
        return headA;
    }
};